# Haskell
Надо установить GHCI (или лучше Haskell Platform)

Пользуемся GHCI, запускаем его, получаем haskell в Интерактивном режиме, вводим выражения `2 + 2`. Запускаем свою прогрумму через:

In [1]:
:load C:\haskell\моя программа.hs
:cd С:\haskell
:load моя программа.hs
:r -- заново загрузить то, что было загружено в прошлый раз

: 

In [2]:
2 + 2

4

Типы данных в Haskell:
```
Int — целые числа, до 2млрд (int в java)
Integer — целые числа произвольной точности (как int python)
Double, Float - дробные числа разной точности, используйте Double
Bool - логические значения True, False
Char - символы 'a', 'c', '%', '#', как в Java
String - строки, "abc", "23dgf", "hello world!"
String = [Char] - список символов
[Int] - список чисел [10, 20, 30, 40]
[Bool] - список Bool [True, False, False]
[Char] - ['a', 'b', 'c', 'd'] эквивалентно "abcd"
````

In [3]:
[10, 20, 30]

[10,20,30]

Можно узнать тип выражения директивой `:type`. Тип в Haskell указывается после `::`

In [5]:
:type ['a', 'b', 'c']

['a', 'b', 'c'] :: [Char]

In [6]:
:type True

True :: Bool

Иногда тип понять не очень просто, даже в обычной ситуации:

In [8]:
:type 2
:type 2.3

2 :: forall p. Num p => p

2.3 :: forall p. Fractional p => p

Это надо понимать так, что у тип `2` это `p`, где p является числовым типом (`Num`). Подходит и Int, и Integer, и Double, и Float.

У `2.3` тип `p`, где `p` является числовым дробным типом (`Fractional`), подходят Double, Float

## Функции

Чтобы ввести функцию, пишем:
`имя аргументы = выражение, которое вычисляет результат`. Обратите внимание, для аргументов функции **скобки не используются**. Скобки нужны для указания порядка вычисления: `(2 + 2) * 2`.

In [12]:
f x = x + 1 -- один аргумент
g x y = x + 2 * y -- два аргумента


Вызвать функцию. Не используем скобки, аргументы пишем через пробел

In [13]:
f 2
g 3 10

3

23

У функций принято подисписывать типы. Это работает как
 * проверка вашей программы, haskell проверяет, что ваше вычисление действительно вычисляет выражение нужного типа
 * документация, т.е. для тех, кто читает, становится проще понять, что делает функция.
 
Часто, если haskell сказал, что тип функции указан правильно, значит, и функция работает правильно.

Тип указывается перед началом функции с помощью `::`

In [18]:
f :: Int -> Int -- Int превращается в Int
f x = x + 1

g :: Int -> Int -> Int -- все три Int как бы равноправны
g x y = x + 2 * y

## Логическое выражение `if`
`if` Bool-условие `then` выражение1 `else` выражение2. if без else не бывает

In [20]:
if 2 == 2 then 10 else 20
if 2 == 3 then 10 else 20

10

20

In [22]:
abs :: Int -> Int
abs x = if x < 0 then -x else x

abs 10
abs 0
abs (-10) -- обязательно скобки, иначе это вычитание abs - 10

10

0

10

При определении функции можно написать несколько равенств, для вычисления будут выбираться те, которые подходят:

In [23]:
h :: Int -> Int
h 0 = 0
h 1 = 2
h x = x + 1

h 10
h 20
h 1
h 0

11

21

2

0

Если ни одно правило не подходит, значит, будет ошибка:

In [25]:
h :: Int -> Int
h 0 = 0
h 1 = 1

h 0
h 1
h 10 -- ошибка. Нет подходящего правила

0

1

: 

В правила можно добавлять условия, они называются guards:

In [27]:
abs :: Int -> Int
abs x | x < 0 = -x  -- правило работает, если выполнено условие
abs x | x >= 0 = x

abs 10
abs 0
abs (-10)

10

0

10

Правила при вычислении функции выполняются по порядку, поэтому проверять x >= 0 не обазятельно:

In [30]:
abs :: Int -> Int
abs x | x < 0 = -x
abs x = x

abs 10

abs :: Int -> Int
abs x | x < 0 = -x
abs x | True = x -- всегда верно, т.е. если не выполнилось x < 0

abs 10

abs :: Int -> Int
abs x | x < 0 = -x
abs x | otherwise = x -- otherwise это True, но так легче читать

abs 10

Line 9: Redundant guard
Found:
abs x | True = x
Why not:
abs x = xLine 15: Redundant guard
Found:
abs x | otherwise = x
Why not:
abs x = x

10

10

10

## Рекурсия

Часто нужно повторять действия, циклов нет, используется вызов функцией самой себя. Например, факториал. `5! = 1 * 2 * 3 * 4 * 5`
Вычисление основано на том, что `5! = 1 * 2 * 3 * 4 * 5 = 4! * 5` и т.д.

In [32]:
fact :: Integer -> Integer
fact 0 = 1
fact 1 = 1 -- не нужная строка, и без нее посчитается
fact n = fact (n - 1) * n

fact 0
fact 1
fact 2
fact 3
fact 4
fact 5

1

1

2

6

24

120

Как происходит вычисление `fact 5`:
```
fact 5 -> fact 4 * 5 -> (fact 3 * 4) * 5 -> ((fact 2 * 3) * 4) * 5 -> (((fact 1 * 2) * 3) * 4) * 5 -> 1 * 2 * 3 * 4 * 5 -> 120
```

**Задача 1**: Написать факториал через if, с одним правилом:
`fact2 x = if ...`

Сделайте файл `tasks1.hs`, создайте там функцию `fact2`. Загрузите этот файл в ghci через `:load task1.hs`, вызывайте `fact2 5`, должно получиться 120.

Неплохо иметь функцию main:

```
main :: IO()
main = do
          print "проверяем fact2"
          print (fact2 10)
          print (fact2 5)
          print (fact2 0)
```

## Списки

Создание списка:
  * перечисление в квадратных скобках через запятую: `[10, 20, 30]` или `[True, False, False]`
  * Операция `:` приписывает элемент в начало списка. `10 : [20, 30]` вычисляется в `[10, 20, 30]`:

In [37]:
10 : [20, 30]
'a' : ['b', 'c', 'd']
10 : 20 : [30, 40, 50] -- сначала приписывает 20, потом 10
10 : 20 : 30 : [] -- приписываем всё к пустому списку

Line 4: Use list literal
Found:
10 : 20 : 30 : []
Why not:
[10, 20, 30]

[10,20,30]

"abcd"

[10,20,30,40,50]

[10,20,30]

Базовые действия со списками это только создание пустого `[]` и операция приписывания к началу `:`. Перечисление в квадратных скобках нужно только для удобства.

### функции на списках
В аргументе функции можно указывать опреации `[]` и `:`. Фактически, при описании функции для аргументов указываются шаблоны (pattern), с которыми надо сопоставлять (match) те аргументы, которые в нее передаются.

In [43]:
list1 :: [Int] -> Int -- функция на списке чисел, выдает число
list1 [10] = 5
list1 [10, 20] = 6
list1 [10, x] = x + 1
list1 [x] = 2 * x
list1 [] = 0
list1 (5:lst) = 5000

list1 [10, 20] -- правило 2 подошло
list1 [10, 30] -- подошло правило 3
list1 [5] -- подходит правило 4
list1 [5, 6, 7, 8] -- подходит правило 6, у нас 5:[6,7,8]
list1 [50, 60, 70, 80] -- подходит правило 6, у нас 5:[6,7,8]

6

31

10

5000

: 

Очень часто функции, которые работают со списками разбирают только два случая, пустой список и не пустой:

In [ ]:
len :: [Int] -> Int -- длина списка
len [] = 0
len (h:t) = 1 + len t

`h` (head) — первый элемент списка, `t` (tail) — оставшаяся часть.
Для `[10, 20, 30]`: `h = 10`, `t = [20, 30]`. `h :: Int`, `t :: [Int]`.

При работе со списком в Haskell встроена только работа с первым элементом. Нельзя напрямую достать, откусить, например, последний элемент.

Еще пример: Узнать первый, второй элемент списка:

In [48]:
get1 :: [Int] -> Int -- такая есть встроенная, называется `head`
get1 (h:t) = h

get2 :: [Int] -> Int
-- get2 (h1:h2:t) = h2
get2 (h:t) = get1 t -- первый элемент хвоста списка

get1 [10, 20, 30]
get2 [10, 20, 30]
get1 []

10

20

: 

Последний пример. Сделать список из указанного числа букв 'x':

In [50]:
repeatX :: Int -> [Char] -- Int -> String
repeatX 0 = []
repeatX n = 'x' : repeatX (n - 1)

repeatX 5

"xxxxx"

Задачи [http://students.iposov.spb.ru/20fall/functional_programming/](http://students.iposov.spb.ru/20fall/functional_programming/). Подсказка

```
fun [] = ...
fun (h:t) = ...
```